### PR TITLE
[DCP] Enable nD device_mesh resharding DTensor in DCP and add associated tests

### PR DESCRIPTION
**Pull request by wz337**
This PR: 
     1. Drop assert for 1D DeviceMesh check to allow DTensor with nD DeviceMesh when creating write_item.
     2. Add tests for both placement changes and mesh changes for both 1D and 2D scenarios. 

cc. @kumpera  @wanchaol  @fegin

### CODE REVIEW

File: test/distributed/checkpoint/test_dtensor_resharding.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+ONE_D_TO_ONE_D_PLACEMENTS = [
+    ([Replicate()], [Shard(0)]),
+    ([Shard(0)], [Replicate()]),
+]
+
+TWO_D_PLACEMENTS = [
+    [Replicate(), Replicate()],
+    [Replicate(), Shard(0)],
+    [Shard(0), Replicate()],
+    [Shard(0), Shard(0)],
```
Comment (kumpera) [inline test/distributed/checkpoint/test_dtensor_resharding.py:39]: Maybe consider adding sharding on non-zero dim?

**Review by kumpera — APPROVED** [review]
LGTM.

File: test/distributed/checkpoint/test_dtensor_resharding.py (hunk 1 above)
Comment (wz337) [reply test/distributed/checkpoint/test_dtensor_resharding.py:39]: Ye. Will add more test cases. Sharding on different dim on placements should be no problem. The previous bug is on offset compute is due to when sharding on one dimension twice, dtensor util only gives the local shard offset instead of global offset. 

Regardless, will definitely add more test cases for this.

File: test/distributed/checkpoint/test_dtensor_resharding.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+]
+TWO_D_TO_TWO_D_PLACEMENTS = []
+for p1 in TWO_D_PLACEMENTS:
+    for p2 in TWO_D_PLACEMENTS:
```
Comment (fduwjj) [inline test/distributed/checkpoint/test_dtensor_resharding.py:43]: nit: you might want to use `itertools.product` instead?

File: test/distributed/checkpoint/test_dtensor_resharding.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            dtensor = distribute_tensor(
+                global_tensor, mesh_1d, placements=placements_1d
+            )
+            state_dict_to_save = {"dtensor": dtensor}
+
+            dist_cp.save_state_dict(
+                state_dict=state_dict_to_save,
+                storage_writer=dist_cp.FileSystemWriter(path=CHECKPOINT_DIR),
+                planner=dist_cp.DefaultSavePlanner(),
+            )
```
Comment (fduwjj) [inline test/distributed/checkpoint/test_dtensor_resharding.py:173]: lots of codes here seems to be duplicate. Maybe you want to consolidate them a little bit?

**Comment by fegin** [comment]
LGTM, @fduwjj's comment is legit. Can improve it in a seperate PR.
